### ISSUE TITLE
Add support to Shared Build Services to receive tasks outputs

### ISSUE THREAD

**Issue by LikeTheSalad** [gradle/gradle#21599]
There is one task, that I believe uses the Worker API to have many instances running in parallel, that I don’t create and don’t have control over, but it has an API that allows me to run my code as part of its execution, so this task is in between Gradle’s APIs and my code, but I need some other tasks to run prior to my code, which is in a Build Service so that it can be reused across all the instances of the task that I don’t control, and so far I haven’t found a way to do so.

### Expected Behavior
It'd be nice if a [Shared Build Service](https://docs.gradle.org/current/userguide/build_services.html) can receive a task’s output as its input, and that’d ensure that said task will always run before the service is created.

### Current Behavior
If I add a task's output as a service's input, the task doesn't run before the service gets created, giving me empty inputs for my service.

### Context

The API I'm using is this one: https://developer.android.com/reference/tools/gradle-api/7.2/com/android/build/api/instrumentation/AsmClassVisitorFactory - It's a factory that allows me to run some code at some point of the android's compilation process, but I'm not sure why even though they have support for parameters, I provide some FileCollections that are generated by other tasks but sometimes it's happened that I get empty inputs, and or an error because apparently my code was trying to access one of them "without waiting for its task to finish running", so maybe there are other issues here, not sure if the Android code is running their task at the right time, or if the parallel running has no way to wait for some tasks to finish for some reason, and since I believe that it should be pretty common for someone running tasks in parallel to use a Build Service as a central piece for it, maybe it'd be useful to allow such services to get managed in the same way as tasks are when it comes to building the graph that decides when should things run and in what order.

**Comment by Vampire** [comment]
Another use-case for a shared build service is to setup and tear down a server needed for testing only when the test task/s is/are going to be executed.
This often involves deploying of a build result that can then be tested by the test code.

Currently, as far as I have figured out, you need to configure an explicit task dependency from the test task to the producer task in Gradle 8.0.2.

If you have the following:
```kts
abstract class Producer : DefaultTask() {
    @get:OutputFile
    abstract val output: RegularFileProperty

    @TaskAction
    fun produce() {
        output.get().asFile.writeText("FOO")
    }
}

val producer by tasks.registering(Producer::class) {
    output.set(layout.buildDirectory.file("foo"))
}

abstract class DeployingService : BuildService<DeployingService.Params> {
    interface Params : BuildServiceParameters {
        val deployable: RegularFileProperty
    }

    @get:InputFile
    val deployable = parameters.deployable
}

abstract class IntegTest : DefaultTask() {
    @get:Nested
    abstract val deployingService: Property<DeployingService>

    @TaskAction
    fun test() {
        println(deployingService.get().deployable.get().asFile.readText())
    }
}

val deployingServiceProvider = gradle.sharedServices.registerIfAbsent("deployingService", DeployingService::class) {
    parameters.deployable.set(producer.flatMap { it.output })
}

val integTest by tasks.registering(IntegTest::class) {
    usesService(deployingServiceProvider)
    deployingService.set(deployingServiceProvider)
}
```
Then it would be nice if calling `gw integTest` would automatically run `producer` first.
Currently if the output file is not present, the task complains as the file is missing.
If the output file is already present, it works and does not give a warning about missing task dependency (will create a separate issue for that).

If you add an explicit `dependsOn(producer)` to the `integTest` task it works as intended, but it would be nice if it worked without.

**Comment by ov7a** [comment]
Sorry for the late reply.

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by abstratt** [comment]
The issue is in the backlog of the relevant team and is prioritized by them.

***

Build services can be used to achieve some of those use cases, but the intended way is to use methods on the services to push/pull data from it. Synchronizing production/consumption can be done via dependencies between tasks.

We could at a minimum validate that build service properties should not be annotated with annotations meant for task properties, and probably that build services parameters do not carry task dependencies.

**Comment by abstratt** [comment]
While we have no plans to implement this feature in a near future, we would like to welcome the community to provide feedback that suggests that a tighter integration between build services and tasks would be required to allow use cases that are not possible currently.

**Comment by Vampire** [comment]
> Synchronizing production/consumption can be done via dependencies between tasks.

How would that be possible?
Given my example, as I said you can add a task dependency from `integTest` to `producer`.
But actually any task that uses the `DeployingService` needs this dependency.
And also only if the `DeployingService` is going to be used because the `integTest` is actually to be run.
With the explicit dependency, `producer` is going to be executed even if `integTest` is not going to be executed and thus the `DeployingService` not going to be started.

Usually, I see and recommend any explicit `dependsOn` that does not have a lifecycle task on the left-hand side as code-smell and bad practice, as it almost always means that instead task outputs should be wired to task inputs which then brings the implicit task dependency as necessary.

**Comment by scottjasso** [comment]
> While we have no plans to implement this feature in a near future, we would like to welcome the community to provide feedback that suggests that a tighter integration between build services and tasks would be required to allow use cases that are not possible currently.

My use case might be an odd one, but:
1) We have a build service that requires some dependencies, so it has a ConfigurableFileCollection parameter set from a ResolvableConfiguration.
2) We have some tasks that use the build service.
3) The tasks have a `@Classpath` input for that build service's FileCollection, so they rerun if the dependencies change.
This works fine, if not a bit odd (seems like that task input should be implied?)

4) We try to use a composite build with dependency substitution (`includeBuild`), to _build_ the BuildService's dependencies rather than downloading them (edit: or, even without an includeBuild -- let's say we wanted the dependencies to come from another subproject). This breaks -- the build service drops all task dependencies, so the tasks don't depend on the includedBuild, so the required dependencies don't exist. There's no good way to add this task dependency as far as I can tell.

The only solution I see is to have _every_ project create a _new_ configuration, to resolve those same dependencies, just so the task can depend on it (and thus depend on the includedBuild tasks). But that seems wrong too -- projects could have different resolution rules and resolve differently than the buildService did, so their task input would be wrong.

**Comment by hfhbd** [comment]
The reason to use task outputs as the input/value of a parameter of a BuildServer is to isolate the classpath. The ([broken](https://github.com/gradle/gradle/issues/25093)) sample from the docs shows a web server but the implementation of the web server itself is an exercise for the reader...

[A build server is a global service](https://github.com/gradle/gradle/issues/27756) and all required classed needs to be part of the buildscript classpath. I hoped to create a setup task with a worker for classpath isolation to isolate my required dependencies from the classpath of the buildscript, but wiring the outputs of the task does not work (this issue).

```kotlin
abstract class TokenPlugin : Plugin<Project> {
    override fun apply(target: Project) {
        val getTokenTask = target.tasks.register("login", Login::class) {
            this.tokenFile.set(target.layout.buildDirectory.file("login.txt"))
        }

        target.gradle.sharedServices.registerIfAbsent("token", TokenBuildService::class) {
            parameters {
                token.set(getTokenTask.flatMap { it.tokenFile })
            }
        }
        target.tasks.register("useToken", ApiTask::class) {

        }
    }
}

abstract class TokenBuildService : BuildService<TokenBuildService.Parameters> {
    interface Parameters : BuildServiceParameters {
        val token: RegularFileProperty
    }

    fun getToken() = parameters.token.asFile.get().readText()
}

abstract class Login: DefaultTask() {

    @get:OutputFile
    abstract val tokenFile: RegularFileProperty

    @TaskAction
    protected fun getToken() {
        tokenFile.get().asFile.writeText("token")
    }
}

abstract class ApiTask: DefaultTask() {
    @get:ServiceReference("token")
    abstract val tokenService: Property<TokenBuildService>

    @TaskAction
    protected fun useToken() {
        println(tokenService.get().getToken())
    }
}

pluginManager.apply(TokenPlugin::class)
```

When running `./gradlew useToken`, the task `login` will not be executed.
